File: typescript/tests/integration/test_agentbay_list_status.test.ts
import { AgentBay, CreateSeesionWithParams } from "../../src/agent-bay";
import { Session } from "../../src/session";
import { log } from "../../src/utils/logger";

describe("AgentBay List Status Integration Tests", () => {
  let agentBay: AgentBay;
  let testSessions: Session[] = [];

  beforeAll(async () => {
    const apiKey = process.env.AGENTBAY_API_KEY;
    if (!apiKey) {
      throw new Error("AGENTBAY_API_KEY environment variable is not set");
    }

    const endpoint = process.env.AGENTBAY_ENDPOINT;
    if (endpoint) {
      agentBay = new AgentBay({ 
        apiKey,
        config: { 
          endpoint,
          timeout_ms: 60000 
        }
      });
      log(`Using endpoint: ${endpoint}`);
    } else {
      agentBay = new AgentBay({ apiKey });
      log("Using default endpoint");
    }
  });

  afterEach(async () => {
    log("\nCleaning up test sessions for this test...");
    for (const session of testSessions) {
      try {
        // Try to resume session first in case it's paused
        try {
          if (session) {
            const result = await agentBay.getStatus(session.sessionId);
            if (result.data?.status === "PAUSED") {
              await session.resumeAsync();
              log(`  ✓ Resumed session: ${session.sessionId}`);
            }
            if (result.data?.status && !["DELETING", "DELETED", "RESUMING", "PAUSING"].includes(result.data.status)) {
              const deleteResult = await agentBay.delete(session);
              if (deleteResult.success) {
                log(`  ✓ Deleted session: ${session.sessionId}`);
              } else {
                log(`  ✗ Failed to delete session: ${session.sessionId}`);
              }
            }
          }
        } catch (resumeError) {
          log(`  ⚠ Could not resume session ${session.sessionId}: ${resumeError}`);
        }
      } catch (e) {
        log(`  ✗ Error deleting session ${session.sessionId}: ${e}`);
      }
    }
    // Clear the list for next test
    testSessions = [];
  });

  const createTestSession = async (): Promise<Session> => {
    const sessionName = `test-pause-resume-${Math.random().toString(36).substring(2, 10)}`;
    log(`\nCreating test session: ${sessionName}`);

    // Create session
    const params: CreateSeesionWithParams = {
      labels: { project: "piaoyun-demo", environment: "testing" },
    };
    
    const result = await agentBay.create(params);
    expect(result.success).toBe(true);
    expect(result.session).toBeDefined();

    const session = result.session!;
    testSessions.push(session);
    log(`  ✓ Session created: ${session.sessionId}`);

    return session;
  };

  const verifySessionStatusAndList = async (
    session: Session, 
    expectedStatuses: string[], 
    operationName = "operation"
  ): Promise<string> => {
    log(`\nVerifying session status after ${operationName}...`);
    
    // First call getStatus to check the current status
    const statusResult = await agentBay.getStatus(session.sessionId);
    expect(statusResult.success).toBe(true);
    expect(statusResult.data).toBeDefined();
    
    const initialStatus = statusResult.data?.status || "UNKNOWN";
    log(`  ✓ Session status from getStatus: ${initialStatus}`);
    expect(expectedStatuses).toContain(initialStatus);

    // Then call getSession for detailed information
    const sessionInfo = await agentBay.getSession(session.sessionId);
    expect(sessionInfo.success).toBe(true);
    expect(sessionInfo.data).toBeDefined();

    const currentStatus = sessionInfo.data?.status || "UNKNOWN";
    expect(currentStatus).toBe(initialStatus);
    log(`  ✓ Session status from getSession: ${currentStatus}`);
    
    // Test list with current status
    const listResult = await agentBay.list({}, 1, 10, currentStatus);
    expect(listResult.success).toBe(true);
    
    // Verify session is in the list and check array structure
    let sessionFound = false;
    for (const sessionData of listResult.sessionIds) {
      if (typeof sessionData === 'object' && sessionData !== null) {
        if ('sessionId' in sessionData && sessionData.sessionId === session.sessionId) {
          sessionFound = true;
          expect(sessionData).toHaveProperty('sessionStatus');
          expect(sessionData).toHaveProperty('sessionId');
          expect((sessionData as any).sessionStatus).toBe(currentStatus);
          break;
        }
      } else {
        log("  ✗ Invalid session data in list result");
        break;
      }
    }
    
    expect(sessionFound).toBe(true);
    log(`  ✓ Session found in list with status ${currentStatus}`);
    log(`  ✓ Session status verification completed for ${operationName}`);
    
    return currentStatus;
  };

  test("should pause and resume session successfully", async () => {
    log("\n" + "=".repeat(60));
    log("TEST: Pause and Resume Session Success");
    log("=".repeat(60));

    // Create a test session
    const session = await createTestSession();

    // Verify session is initially in RUNNING state
    const statusResult = await agentBay.getStatus(session.sessionId);
    expect(statusResult.success).toBe(true);
    expect(statusResult.data).toBeDefined();
    
    const initialStatus = statusResult.data?.status || "UNKNOWN";
    log(`  ✓ Session status from getStatus: ${initialStatus}`);
    expect(initialStatus).toBe("RUNNING");

    // Pause the session
    log(`\nStep 2: Pausing session...`);
    const pauseResult = await session.pauseAsync();

    // Verify pause result
    expect(pauseResult.success).toBe(true);
    log(`  ✓ Session pause initiated successfully`);
    log(`    Request ID: ${pauseResult.requestId}`);

    // Wait a bit for pause to complete
    log(`\nStep 3: Waiting for session to pause...`);
    await new Promise(resolve => setTimeout(resolve, 2000));

    // Verify session status after pause
    await verifySessionStatusAndList(session, ["PAUSED", "PAUSING"], "pause");
  }, 120000);

  test("should resume session successfully", async () => {
    log("\n" + "=".repeat(60));
    log("TEST: Resume Session Success");
    log("=".repeat(60));

    // Create a test session
    const session = await createTestSession();

    // Pause the session first
    log(`\nStep 1: Pausing session...`);
    const pauseResult = await session.pauseAsync();
    expect(pauseResult.success).toBe(true);
    log(`  ✓ Session pause initiated successfully`);

    // Wait for pause to complete
    log(`\nStep 2: Waiting for session to pause...`);
    await new Promise(resolve => setTimeout(resolve, 2000));

    // Session should be PAUSED or PAUSING after pause operation
    const statusResult = await agentBay.getStatus(session.sessionId);
    expect(statusResult.success).toBe(true);
    expect(statusResult.data).toBeDefined();
    
    const initialStatus = statusResult.data?.status || "UNKNOWN";
    log(`  ✓ Session status from getStatus: ${initialStatus}`);
    expect(["PAUSED", "PAUSING"]).toContain(initialStatus);
    log(`  ✓ Session status checked`);

    const resumeResult = await session.resumeAsync();

    // Verify resume result
    expect(resumeResult.success).toBe(true);
    log(`  ✓ Session resume initiated successfully`);
    log(`    Request ID: ${resumeResult.requestId}`);

    // Wait a bit for resume to complete
    log(`\nStep 4: Waiting for session to resume...`);
    await new Promise(resolve => setTimeout(resolve, 2000));

    // Verify session status after resume
    await verifySessionStatusAndList(session, ["RUNNING", "RESUMING"], "resume");
  }, 120000);

  test("should pause and delete session successfully", async () => {
    log("\n" + "=".repeat(60));
    log("TEST: Pause and Delete Session Success");
    log("=".repeat(60));
    
    log(`\nStep 1: Creating test session...`);
    // Create a test session
    const session = await createTestSession();
    
    // Pause the session
    log(`\nStep 2: Pausing session...`);
    const pauseResult = await session.pauseAsync();

    // Verify pause result
    expect(pauseResult.success).toBe(true);
    log(`  ✓ Session pause initiated successfully`);
    log(`    Request ID: ${pauseResult.requestId}`);

    // Wait a bit for pause to complete
    log(`\nStep 3: Waiting for session to pause...`);
    await new Promise(resolve => setTimeout(resolve, 2000));

    log(`  ✓ Session status after pause checked`);
    
    // Delete the session
    log(`\nStep 4: Deleting session...`);
    const deleteResult = await agentBay.delete(session);
    if (deleteResult.success) {
      log("delete session successfully");
    }

    // Verify session status after delete
    await verifySessionStatusAndList(session, ["DELETING", "DELETED"], "delete");
  }, 120000);

  test("should list sessions with status filter", async () => {
    log("\n" + "=".repeat(60));
    log("TEST: List Sessions with Status Filter");
    log("=".repeat(60));

    // Create a test session
    const session = await createTestSession();

    // Test listing with RUNNING status
    log(`\nTesting list with RUNNING status filter...`);
    const runningListResult = await agentBay.list({}, 1, 10, "RUNNING");
    expect(runningListResult.success).toBe(true);
    
    // Verify our session is in the RUNNING list
    const runningSessionFound = runningListResult.sessionIds.some(sessionData => {
      if (typeof sessionData === 'object' && sessionData !== null && 'sessionId' in sessionData) {
        return (sessionData as any).sessionId === session.sessionId;
      }
      return false;
    });
    expect(runningSessionFound).toBe(true);
    log(`  ✓ Session found in RUNNING status list`);

    // Test listing with invalid status
    log(`\nTesting list with invalid status filter...`);
    const invalidListResult = await agentBay.list({}, 1, 10, "INVALID_STATUS");
    expect(invalidListResult.success).toBe(false);
    expect(invalidListResult.errorMessage).toContain("Invalid status");
    log(`  ✓ Invalid status correctly rejected`);

    // Test listing without status filter
    log(`\nTesting list without status filter...`);
    const allListResult = await agentBay.list();
    expect(allListResult.success).toBe(true);
    expect(allListResult.sessionIds.length).toBeGreaterThan(0);
    log(`  ✓ List without filter returned ${allListResult.sessionIds.length} sessions`);
  }, 60000);

  test("should handle pagination correctly", async () => {
    log("\n" + "=".repeat(60));
    log("TEST: Pagination Handling");
    log("=".repeat(60));

    // Test first page
    const page1Result = await agentBay.list({}, 1, 5);
    expect(page1Result.success).toBe(true);
    log(`  ✓ Page 1 returned ${page1Result.sessionIds.length} sessions`);

    // Test invalid page number
    const invalidPageResult = await agentBay.list({}, 0, 5);
    expect(invalidPageResult.success).toBe(false);
    expect(invalidPageResult.errorMessage).toContain("Page number must be >= 1");
    log(`  ✓ Invalid page number correctly rejected`);
  }, 30000);
});

File: golang/tests/pkg/integration/test_agentbay_list_status_test.go
package integration

import (
	"fmt"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestAgentBayListStatusIntegration tests session pause and resume operations with list status verification
func TestAgentBayListStatusIntegration(t *testing.T) {
	// Get API Key
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}
	fmt.Printf("api_key = %s\n", apiKey)

	// Initialize AgentBay client
	agentBay, err := agentbay.NewAgentBay(apiKey, nil)
	require.NoError(t, err)
	require.NotNil(t, agentBay)

	// Create session for testing
	fmt.Println("Creating a new session for pause/resume testing...")
	params := agentbay.NewCreateSessionParams().
		WithLabels(map[string]string{
			"project":     "piaoyun-demo",
			"environment": "testing",
		})

	fmt.Println("Creating session...")
	sessionResult, err := agentBay.Create(params)
	require.NoError(t, err)
	require.NotNil(t, sessionResult)
	require.True(t, sessionResult.Success)
	require.NotNil(t, sessionResult.Session)

	session := sessionResult.Session
	fmt.Printf("Session created with ID: %s\n", session.SessionID)

	// Helper function to verify session status and list
	verifySessionStatusAndList := func(expectedStatuses []agentbay.SessionStatus, operationName string) agentbay.SessionStatus {
		fmt.Printf("\nVerifying session status after %s...\n", operationName)

		// First call GetStatus to check the current status
		statusResult, err := agentBay.GetStatus(session.SessionID)
		require.NoError(t, err)
		require.True(t, statusResult.Success, fmt.Sprintf("Failed to get session status: %s", statusResult.ErrorMessage))

		initialStatus := "UNKNOWN"
		if statusResult.Data != nil {
			initialStatus = statusResult.Data.Status
		}
		fmt.Printf("  ✓ Session status from GetStatus: %s\n", initialStatus)

		// Verify status is in expected list
		found := false
		for _, expected := range expectedStatuses {
			if initialStatus == expected.String() {
				found = true
				break
			}
		}
		require.True(t, found, fmt.Sprintf("Unexpected status %s, expected one of %v", initialStatus, expectedStatuses))

		// GetSession is internal in SDK; use GetStatus only.
		currentStatus := initialStatus

		// Test list with current status
		listResult, err := agentBay.List(currentStatus, nil, nil, nil)
		require.NoError(t, err)

		// Verify session is in the list and check array structure
		sessionFound := false
		for _, sessionData := range listResult.SessionIds {
			if sessionID, exists := sessionData["sessionId"]; exists && sessionID == session.SessionID {
				sessionFound = true
				require.Contains(t, sessionData, "sessionStatus", "sessionStatus field missing in list result")
				require.Contains(t, sessionData, "sessionId", "sessionId field missing in list result")
				require.Equal(t, currentStatus, sessionData["sessionStatus"])
				break
			}
		}

		require.True(t, sessionFound, fmt.Sprintf("Session %s not found in list with status %s", session.SessionID, currentStatus))
		fmt.Printf("  ✓ Session found in list with status %s\n", currentStatus)
		fmt.Printf("  ✓ Session status verification completed for %s\n", operationName)

		return agentbay.SessionStatus(currentStatus)
	}

	// Ensure cleanup
	defer func() {
		fmt.Println("\nCleaning up test sessions for this test...")
		if session != nil {
			// Try to resume session first in case it's paused
			statusResult, err := agentBay.GetStatus(session.SessionID)
			if err != nil {
				fmt.Printf("  ⚠ Could not get session status %s: %v\n", session.SessionID, err)
			} else if statusResult.Success && statusResult.Data != nil {
				currentStatus := statusResult.Data.Status
				
				// Resume if paused
				if currentStatus == "PAUSED" {
					_, resumeErr := agentBay.Resume(session, 600, 2.0)
					if resumeErr != nil {
						fmt.Printf("  ⚠ Could not resume session %s: %v\n", session.SessionID, resumeErr)
					} else {
						fmt.Printf("  ✓ Resumed session: %s\n", session.SessionID)
					}
				}
				
				// Delete if not already deleting or deleted
				if currentStatus != "DELETING" && currentStatus != "DELETED" && 
				   currentStatus != "RESUMING" && currentStatus != "PAUSING" {
					deleteResult, err := agentBay.Delete(session, false)
					if err != nil {
						fmt.Printf("  ✗ Error deleting session %s: %v\n", session.SessionID, err)
					} else if deleteResult != nil {
						if deleteResult.Success {
							fmt.Printf("  ✓ Deleted session: %s\n", session.SessionID)
						} else {
							fmt.Printf("  ✗ Failed to delete session: %s\n", session.SessionID)
						}
					}
				}
			}
		}
	}()

	// Test 1: Pause and Resume Session Success
	t.Run("Pause and Resume Session Success", func(t *testing.T) {
		fmt.Println("\n" + strings.Repeat("=", 60))
		fmt.Println("TEST: Pause and Resume Session Success")
		fmt.Println(strings.Repeat("=", 60))

		// Verify session is initially in RUNNING state
		statusResult, err := agentBay.GetStatus(session.SessionID)
		require.NoError(t, err)
		require.True(t, statusResult.Success, fmt.Sprintf("Failed to get session status: %s", statusResult.ErrorMessage))

		initialStatus := "UNKNOWN"
		if statusResult.Data != nil {
			initialStatus = statusResult.Data.Status
		}
		fmt.Printf("  ✓ Session status from GetStatus: %s\n", initialStatus)
		require.Equal(t, "RUNNING", initialStatus, fmt.Sprintf("Unexpected status %s, expected RUNNING", initialStatus))

		// Pause the session
		fmt.Printf("\nStep 2: Pausing session...\n")
		pauseResult, err := agentBay.Pause(session, 600, 2.0)
		require.NoError(t, err)
		require.NotNil(t, pauseResult)
		require.True(t, pauseResult.Success, fmt.Sprintf("Pause failed: %s", pauseResult.ErrorMessage))
		fmt.Printf("  ✓ Session pause initiated successfully\n")
		fmt.Printf("    Request ID: %s\n", pauseResult.RequestID)

		// Wait a bit for pause to complete
		fmt.Printf("\nStep 3: Waiting for session to pause...\n")
		time.Sleep(2 * time.Second)

		// Verify session status after pause
		verifySessionStatusAndList([]agentbay.SessionStatus{agentbay.SessionStatusPaused, agentbay.SessionStatusPausing}, "pause")
	})

	// Test 2: Resume Session Success
	t.Run("Resume Session Success", func(t *testing.T) {
		fmt.Println("\n" + strings.Repeat("=", 60))
		fmt.Println("TEST: Resume Session Success")
		fmt.Println(strings.Repeat("=", 60))

		// Session should be PAUSED or PAUSING after previous test
		statusResult, err := agentBay.GetStatus(session.SessionID)
		require.NoError(t, err)
		require.True(t, statusResult.Success, fmt.Sprintf("Failed to get session status: %s", statusResult.ErrorMessage))

		initialStatus := "UNKNOWN"
		if statusResult.Data != nil {
			initialStatus = statusResult.Data.Status
		}
		fmt.Printf("  ✓ Session status from GetStatus: %s\n", initialStatus)
		require.Contains(t, []string{"PAUSED", "PAUSING"}, initialStatus, fmt.Sprintf("Unexpected status %s, expected PAUSED or PAUSING", initialStatus))

		// Resume the session
		fmt.Printf("\nStep 1: Resuming session...\n")
		resumeResult, err := agentBay.Resume(session, 600, 2.0)
		require.NoError(t, err)
		require.NotNil(t, resumeResult)
		require.True(t, resumeResult.Success, fmt.Sprintf("Resume failed: %s", resumeResult.ErrorMessage))
		fmt.Printf("  ✓ Session resume initiated successfully\n")
		fmt.Printf("    Request ID: %s\n", resumeResult.RequestID)

		// Wait a bit for resume to complete
		fmt.Printf("\nStep 2: Waiting for session to resume...\n")
		time.Sleep(2 * time.Second)

		// Verify session status after resume
		verifySessionStatusAndList([]agentbay.SessionStatus{agentbay.SessionStatusRunning, agentbay.SessionStatusResuming}, "resume")
	})

	// Test 3: List with Status Filter
	t.Run("List with Status Filter", func(t *testing.T) {
		fmt.Println("\n" + strings.Repeat("=", 60))
		fmt.Println("TEST: List with Status Filter")
		fmt.Println(strings.Repeat("=", 60))

		// Test listing with RUNNING status filter
		fmt.Printf("Testing list with RUNNING status filter...\n")
		listResult, err := agentBay.List(agentbay.SessionStatusRunning.String(), nil, nil, nil)
		require.NoError(t, err)

		// Verify our session is in the RUNNING list
		sessionFound := false
		for _, sessionData := range listResult.SessionIds {
			if sessionID, exists := sessionData["sessionId"]; exists && sessionID == session.SessionID {
				sessionFound = true
				require.Equal(t, "RUNNING", sessionData["sessionStatus"])
				break
			}
		}
		require.True(t, sessionFound, "Session should be found in RUNNING status list")
		fmt.Printf("  ✓ Session found in RUNNING status list\n")

		// Test using enum with ListByStatus method
		fmt.Printf("Testing ListByStatus with enum...\n")
		_, err = agentBay.ListByStatus(agentbay.SessionStatusRunning, nil, nil, nil)
		require.NoError(t, err)
		fmt.Printf("  ✓ ListByStatus with enum works correctly\n")

		// Test invalid status validation
		fmt.Printf("Testing invalid status validation...\n")
		_, err = agentBay.List("INVALID_STATUS", nil, nil, nil)
		require.Error(t, err)
		require.Contains(t, err.Error(), "invalid session status 'INVALID_STATUS'")
		fmt.Printf("  ✓ Invalid status validation works correctly\n")
	})

	fmt.Println("\nTest completed. Session will be kept alive for 5 seconds before deletion.")
	fmt.Println("You can inspect the session during this time.")
	time.Sleep(5 * time.Second)
}

// TestAgentBaySessionStatusEnumFunctionality tests the SessionStatus enum functionality
func TestAgentBaySessionStatusEnumFunctionality(t *testing.T) {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("TEST: AgentBay SessionStatus Enum Functionality")
	fmt.Println(strings.Repeat("=", 60))

	// Test valid statuses
	validStatuses := []agentbay.SessionStatus{
		agentbay.SessionStatusRunning,
		agentbay.SessionStatusPaused,
		agentbay.SessionStatusPausing,
		agentbay.SessionStatusResuming,
		agentbay.SessionStatusDeleted,
		agentbay.SessionStatusDeleting,
		agentbay.SessionStatusUnknown,
	}

	for _, status := range validStatuses {
		t.Run(fmt.Sprintf("Valid status %s", status), func(t *testing.T) {
			assert.True(t, status.IsValid(), fmt.Sprintf("Status %s should be valid", status))
			assert.NotEmpty(t, status.String(), fmt.Sprintf("Status %s string should not be empty", status))
		})
	}

	// Test invalid status
	t.Run("Invalid status", func(t *testing.T) {
		invalidStatus := agentbay.SessionStatus("INVALID")
		assert.False(t, invalidStatus.IsValid(), "Invalid status should return false")
	})

	// Test GetValidStatuses function
	t.Run("GetValidStatuses function", func(t *testing.T) {
		allValidStatuses := agentbay.GetValidStatuses()
		assert.Len(t, allValidStatuses, 7, "Should return 7 valid statuses")
		
		for _, status := range allValidStatuses {
			assert.True(t, status.IsValid(), fmt.Sprintf("Status %s from GetValidStatuses should be valid", status))
		}
	})

	fmt.Println("SessionStatus enum tests completed successfully")
}
